channeldb+contractcourt: add resolved resolution reports to the breached channel close summary

bjarnemagnussen opened the pull request:
## Change Description
Currently a breached channel close will not add the actual resolution reports to the summary. This PR adds those reports by making use of the possible communication that already exists between the breacharbiter and the breach resolver.

I know that there is a todo-item to actually refactor parts of the breacharbiter into the breach resolver, which more naturally would allow to also resolution reports to the channel close summary. But as this may require more work, time and scrutiny, I hope that this PR will be considered for the time being.

## Steps to Test
With these changes when a channel is breached it will show the resolution reports in the close summary.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

Crypt-iQ commented on contractcourt/breacharbiter.go:
```diff
@@ -643,6 +652,29 @@ func updateBreachInfo(breachInfo *retributionInfo, spends []spend) (
 		doneOutputs[s.index] = struct{}{}
+
+		// Send report of the resolved output if there is a subscription.
+		if n, ok := b.subscriptions[breachInfo.chanPoint]; ok {
```
Instead of this, I think it would be more elegant to append to a slice stored in `b.subscriptions` and send them all at the end in `notifyBreachComplete` and avoid the wait groups, goroutines. Then the breachresolver doesn't have to wait for a closed channel and can instead wait to receive the reports in a batch rather than incrementally. If for some reason the reports are empty, that should be handled also.

Crypt-iQ replied:
It is also possible here that the counter-party manages to spend their breached outputs, whether it's their to_local output or their second-level htlc outputs. It would be rare as lnd would need to be down, borked, etc, but then the accounting would be incorrect which would be the least of your worries. I think adding logic here would be a nice future improvement, maybe add a comment?

Crypt-iQ reviewed (commented):
This pr is mostly there, I think it could use a unit test to assert that reports are sent. Another test could check that even if the breacharbiter goes down before completion and fails to send the batch of reports (per my suggestion below), on restart it will send the exact same batch.

lightninglabs-deploy commented:
@carlakc: review reminder
@bjarnemagnussen, remember to re-request review from reviewers when ready

bjarnemagnussen commented:
Thanks @Crypt-iQ for your review and helpful comments!

I have addressed the issue of not checking if we actually are the peer that spends from a breached output. I have repurposed the existing `channeldb.ResolverOutcome` as follows, where for each we use `channeldb.ResolverTypeBreached` in the reporting:
* `channeldb.ResolverOutcomeClaimed`: We claimed the breached output
* `channeldb.ResolverOutcomeTimeout`: The remote spent the output as we did not react within our time limit
* `channeldb.ResolverOutcomeFirstLevel`: The remote did spend the HTLC to reach the second-level, which we hopefully claim with another resolution.

I have also added unit tests for these cases. 

However, I did not add a test to check for the case that the breach arbiter is shutdown before resolving the breach, as I believe this would actually involve a integration test due to the spend notification being "replayed"?

lightninglabs-deploy commented:
@crypt-iq: review reminder
@carlakc: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder
@carlakc: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder
@carlakc: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
Closing due to inactivity

lightninglabs-deploy commented:
Closing due to inactivity

lightninglabs-deploy commented:
Closing due to inactivity
